### PR TITLE
redis cache - better parsing of connection uri

### LINKED ISSUE #497: redis cache plugin does not accept (bracketed) IPv6 addresses
##### SUMMARY

The redis cache plugin does not accept a fact_caching_connection with an IPv6 address

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

redis cache plugin

##### ANSIBLE VERSION

```paste below
ansible 2.9.6
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/ansible/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.3 (default, Dec 20 2019, 18:57:59) [GCC 8.3.0]
```

##### CONFIGURATION

```paste below
ANSIBLE_PIPELINING(/etc/ansible/ansible.cfg) = True
INTERPRETER_PYTHON(/etc/ansible/ansible.cfg) = auto_legacy_silent
RETRY_FILES_ENABLED(/etc/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT

I used Debian Buster 10.4 backports packages

##### STEPS TO REPRODUCE

Change CACHE_PLUGIN_CONNECTION(/etc/ansible/ansible.cfg) to a value with an IPv6 address, e.g. "[::1]:6379:0" (In my actual use case, its a link local unicast address from another machine, but ip6-localhost produces the same result) and use the redis fact cache plugin

```yaml
- hosts: all
  tasks:
    - setup:
```

##### EXPECTED RESULTS

The play would run

##### ACTUAL RESULTS

ansible won't run at all

```paste below
ERROR! Unexpected Exception, this is probably a bug: invalid literal for int() with base 10: ''
the full traceback was:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/redis/connection.py", line 988, in get_connection
    connection = self._available_connections.pop()
IndexError: pop from empty list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/ansible", line 123, in <module>
    exit_code = cli.run()
  File "/usr/lib/python3/dist-packages/ansible/cli/adhoc.py", line 165, in run
    result = self._tqm.run(play)
  File "/usr/lib/python3/dist-packages/ansible/executor/task_queue_manager.py", line 241, in run
    play_return = strategy.run(iterator, play_context)
  File "/usr/lib/python3/dist-packages/ansible/plugins/strategy/linear.py", line 282, in run
    _hosts=self._hosts_cache, _hosts_all=self._hosts_cache_all)
  File "/usr/lib/python3/dist-packages/ansible/vars/manager.py", line 305, in get_vars
    facts = wrap_var(self._fact_cache.get(host.name, {}))
  File "/usr/lib/python3.7/_collections_abc.py", line 660, in get
    return self[key]
  File "/usr/lib/python3/dist-packages/ansible/vars/fact_cache.py", line 31, in __getitem__
    if not self._plugin.contains(key):
  File "/usr/lib/python3/dist-packages/ansible/plugins/cache/redis.py", line 132, in contains
    self._expire_keys()
  File "/usr/lib/python3/dist-packages/ansible/plugins/cache/redis.py", line 125, in _expire_keys
    self._db.zremrangebyscore(self._keys_set, 0, expiry_age)
  File "/usr/lib/python3/dist-packages/redis/client.py", line 2532, in zremrangebyscore
    return self.execute_command('ZREMRANGEBYSCORE', name, min, max)
  File "/usr/lib/python3/dist-packages/redis/client.py", line 772, in execute_command
    connection = pool.get_connection(command_name, **options)
  File "/usr/lib/python3/dist-packages/redis/connection.py", line 990, in get_connection
    connection = self.make_connection()
  File "/usr/lib/python3/dist-packages/redis/connection.py", line 1025, in make_connection
    return self.connection_class(**self.connection_kwargs)
  File "/usr/lib/python3/dist-packages/redis/connection.py", line 452, in __init__
    self.port = int(port)
ValueError: invalid literal for int() with base 10: ''

```

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY
Improved the parsing of the connection string, from `.split(':')` to a regexp.

Fixes #497 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins/cache/redis.py

### CODE REVIEW

File: tests/unit/plugins/cache/test_redis.py [hunk 1]
```diff
@@ -19,14 +19,31 @@
...
+import os
+
 import pytest
 
 pytest.importorskip('redis')
 
 from ansible.plugins.loader import cache_loader
 from ansible_collections.community.general.plugins.cache.redis import CacheModule as RedisCache
+from ansible import constants as C
+from ansible.release import __version__ as av
```
Comment (felixfontein) [inline tests/unit/plugins/cache/test_redis.py:31]:
```suggestion
from ansible.release import __version__ as ansible_version
```

File: tests/unit/plugins/cache/test_redis.py [hunk 2]
```diff
@@ -19,14 +19,31 @@
...
+
+version = tuple(int(x) for x in av.split('.')[0:2])
+
+
+def test_redis_cachemodule():
+    # The _uri option is required for the redis plugin
+    connection = 'ni 127.0.0.1:6379:1'
+    if version == (2, 9):
```
Comment (felixfontein) [inline tests/unit/plugins/cache/test_redis.py:39]: Why not just:
```suggestion
    if ansible_version.startswith('2.9.'):
```

File: tests/unit/plugins/cache/test_redis.py [hunk 3]
```diff
@@ -19,14 +19,31 @@
 from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
 
+import os
```
Comment (felixfontein) [inline tests/unit/plugins/cache/test_redis.py:22]: What's this needed for?

Comment (russoz) [reply tests/unit/plugins/cache/test_redis.py]: Oh, leftover from something I was testing. Removed.

File: tests/unit/plugins/cache/test_redis.py [hunk 4]
```diff
@@ -25,8 +25,21 @@
...
+from ansible import constants as C
+from ansible.release import __version__ as ansible_version
 
 
 def test_redis_cachemodule():
     # The _uri option is required for the redis plugin
-    assert isinstance(cache_loader.get('community.general.redis', **{'_uri': '127.0.0.1:6379:1'}), RedisCache)
+    connection = 'ni 127.0.0.1:6379:1'
+    if ansible_version.startswith('2.9.'):
+        C.CACHE_PLUGIN_CONNECTION = "c." + connection
```
Comment (felixfontein) [inline tests/unit/plugins/cache/test_redis.py:36]:
```suggestion
        C.CACHE_PLUGIN_CONNECTION = connection
```
What's the `c.` doing there?

Comment (felixfontein) [reply tests/unit/plugins/cache/test_redis.py:36]: (Same question for the `ni ` above.)

Comment (russoz) [reply tests/unit/plugins/cache/test_redis.py]: I had a test case in the company's code with Rammstein's lyrics once. :-)

Comment (russoz) [reply tests/unit/plugins/cache/test_redis.py]: Removed

File: tests/unit/plugins/cache/test_redis.py [hunk 5]
```diff
@@ -25,8 +25,21 @@
 
 from ansible.plugins.loader import cache_loader
 from ansible_collections.community.general.plugins.cache.redis import CacheModule as RedisCache
+from ansible import constants as C
+from ansible.release import __version__ as ansible_version
```
Comment (felixfontein) [inline tests/unit/plugins/cache/test_redis.py:29]: How about ordering the imports? :)
```suggestion
from ansible import constants as C
from ansible.plugins.loader import cache_loader
from ansible.release import __version__ as ansible_version
from ansible_collections.community.general.plugins.cache.redis import CacheModule as RedisCache
```

File: plugins/cache/redis.py [hunk 6]
```diff
@@ -91,6 +92,8 @@ class CacheModule(BaseCacheModule):
     performance.
     """
     _sentinel_service_name = None
+    re_url_conn = re.compile(r'^(.*):(\d+):(\d+)(?::(.*))?$')
```
Comment (felixfontein) [inline plugins/cache/redis.py:95]: This regex can split `1:2:3:4` as `'1:2'`, `'3'`, `'4'`, `''`. This is not what you intend.

How about using the following instead:
```suggestion
    re_url_conn = re.compile(r'^([^:]+|\[[^]]+\]):(\d+):(\d+)(?::(.*))?$')
```

File: plugins/cache/redis.py [hunk 7]
```diff
@@ -91,6 +92,8 @@ class CacheModule(BaseCacheModule):
...
+    re_sent_conn = re.compile(r'(.*):(\d+)')
```
Comment (felixfontein) [inline plugins/cache/redis.py:96]:
```suggestion
    re_sent_conn = re.compile(r'^(.*):(\d+)$')
```

**Review by felixfontein — COMMENTED** [review]
Looks good to me. The tests should really be improved at some point, right now they don't really test much. But that's not caused by this PR :)

**Comment by felixfontein** [comment]
I'll merge in a few days if nobody complains.

**Comment by felixfontein** [comment]
@russoz thanks for improving this!

**Pull request merged**